### ISSUE TITLE
Adapt (or rewrite) semantic-release to manage release process.

### ISSUE THREAD

**Issue by kevingranade** [CleverRaven/Cataclysm-DDA#21325]
Ran across https://github.com/semantic-release/semantic-release

It doesn't do what we need (it wants to publish a release with changelog with every commit AFAICT, and it'd deeply tied to NPM), but what it DOES do is provide a framework for automatic generation of changelogs based on commit messages.  This in turn requires commit messages to follow a certain format to be parseable.  In short you insert tags into your commit message for later extraction, e.g. "Fix: zombie staggering", "Content: Added yet another AK variant".

This kind of formatting wasn't previously viable because it was opt-in by the commit author, and there was no easy enforcement method.  Now however, we're integrated with Travis, and we can make assertions about commit messages such as "each PR must have at least one commit that specifies a changelog entry", which is enforced by something like https://github.com/conventional-changelog/validate-commit-msg

With this enforcement in place, specifying the changelog entry for your PR just becomes part of the PR review process, and at release time it's a simple matter of either running the changelog creation script locally and then lightly editing it, or possibly just pushing a release tag with a header.

**Comment by kevingranade** [comment]
I was mistaken about some details, it looked like semantic-release wanted to release with every commit, but instead it wants to release with every push to a particular branch, which we could definitely do.

**Comment by kevingranade** [comment]
Side note, this should also be able to build an incremental changelog for the development branch, satisfying the frequent request for more visibility into the development process.

**Comment by Leland** [comment]
Yup, you nailed it. I'm all for this – semantic versioning is how I've approached development in most every project.

With CDDA, it would mean a change to the repository so that there's an "experimental" or "development" branch that PRs are merged into. Semver would be incremented every time that a merge is made from the development branch into the master branch.

In other words, the master branch being updated and a release being created would become synonymous. 

The commit messages would be nice, haven't made use of those before. I'm sure that having someone actually write out release notes would get tiring quickly if there's a planned move to more frequent releases.

**Comment by kevingranade** [comment]
Alternately just add a release branch, merging to it would trigger a release, done.
Unfortunately, changelog effort scales with number of commits, not release frequency.  When we call the current release ready I'm going to spend probably a week just drafting the changelog :/

**Comment by kevingranade** [comment]
Slight update, recommended package for commit linting is:
https://github.com/marionebl/commitlint
And the recommended package for creating changelogs is:
http://conventionalcommits.org/

Alternately, github-changelog-generator is a lighter weight solution, but probably needs a linter for pull requests to make sure it can incorporate the entries.
https://github.com/skywinder/github-changelog-generator/blob/master/README.md

**Comment by AMurkin** [comment]
Can we have commitlint available via web browser similar to JSON linter/formatter?

**Comment by kevingranade** [comment]
I don't know, there will be a commitlint config available for local
checking, don't know offhand if it will be simple to stand up a web form
for it.

**Comment by kevingranade** [comment]
One point of weirdness, it looks like conventional commits is fine with commits not being "well-formed", they're just ignored.
commitlint however wants every commit to be well-formed, which pushes you into a workflow where each changelog entry corresponds to one commit, and every commit must be "well-formed".
I think I can work around this by writting a wrapper script that calls the commitlint API, then it can hopefully do something like, "Retrieve commits for a pull request, verify that at least one commit in the pull request contains a changelog entry".

If not that, there's a completely different approach based on pull requests that would lint the pull request messages instead of the commit messages, and build the changelog from that.
